scipy/_lib/_uarray/_uarray_dispatch.cxx

```diff
@@ -15,13 +15,43 @@
 
 namespace {
 
+template <typename T>
+class immortal {
+    alignas(T) std::byte storage[sizeof(T)];
+
+public:
+    template <typename... Args>
+    immortal(Args&&... args) {
+       // Construct new T in storage
+       new(&storage) T(std::forward<Args>(args)...);
+    }
+    ~immortal() {
+        // Intentionally don't call destructor
+    }
+
+    // T* get() const { return reinterpret_cast<T*>(&storage); }
+    const T* get() const { return reinterpret_cast<const T*>(&storage); }
+    const T* get_const() const { return reinterpret_cast<const T*>(&storage); }
+
+
+    const T* operator ->() const { return get(); }
+    // T* operator ->() const { return get(); }
+
+    friend bool operator==(const immortal<T> & lhs, const immortal<T> & rhs) {
+      return lhs.get() == rhs.get();
+    }
+
+};
+
 /** Handle to a python object that automatically DECREFs */
 class py_ref {
-  explicit py_ref(PyObject * object): obj_(object) {}
+  // explicit py_ref(PyObject * object): obj_(object) {}
 
 public:
   py_ref() noexcept: obj_(nullptr) {}
+  py_ref(PyObject * object): obj_(object) {}
   py_ref(std::nullptr_t) noexcept: py_ref() {}
+  py_ref(immortal<py_ref> other): obj_(other->get()) { Py_XINCREF(obj_); }
```

Comment:
These should not be necessary. `py_ref` should only ever be created via `steal` or `ref`.